src/NUnitFramework/framework/Attributes/CsvDataSourceAttribute.cs

```diff
@@ -0,0 +1,206 @@
+// ***********************************************************************
+
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Xml;
+using System.Collections;
+using System.Reflection;
+using NUnit.Framework.Interfaces;
+using NUnit.Framework.Internal;
+using NUnit.Framework.Internal.Builders;
+
+
+namespace NUnit.Framework.Attributes
+{
+    /// <summary>
+    /// Specifies an external data source for test methods. This is evaulated at compile time.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true, Inherited = false)]
+    public class CsvDataAttribute : NUnitAttribute, ITestBuilder, IImplyFixture
+    {
+        private NUnitTestCaseBuilder _builder = new NUnitTestCaseBuilder();
+        private CsvData csv;
+
+        #region Constructors
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="sourceName">path of data file</param>
+        /// <param name="rowCount"> rows to read. Defaults to -1 for all rows</param>
+        public CsvDataAttribute(string sourceName, int rowCount = -1)
+        {
+            if (rowCount == -1)
+            {
+                csv = new CsvData(sourceName);
+                
+            }
+            else {
+                csv = new CsvData(sourceName, rowCount);
+            }
+        }
+
+        #endregion
+
+
+        #region ITestCaseSource Members
+
+        /// <summary>
+        /// Returns a set of test case parameters for use as arguments
+        /// to a parameterized test method.
+        /// </summary>
+        /// <param name="method">The method for which data is needed.</param>
+        /// <returns></returns>
+        private IEnumerable<TestCaseParameters> GetTestCasesFor(IMethodInfo method)
+        {
+            List<TestCaseParameters> data = new List<TestCaseParameters>();
+
+            try
+            {
+                IEnumerable source = GetDataFromSource(method);
+
+                if (source != null)
+                {
+#if NETCF
+              ParameterInfo[] parameters = method.IsGenericMethodDefinition ? new ParameterInfo[0] : method.GetParameters();
+#else
+                    IParameterInfo[] parameters = method.GetParameters();
+#endif
+
+                    foreach (object item in source)
+                    {
+                        var parms = item as TestCaseParameters;
+
+                        if (parms == null)
+                        {
+                            object[] args = item as object[];
+                            if (args != null)
+                            {
+#if NETCF
+                    if (method.IsGenericMethodDefinition)
+                    {
+                      var mi = method.MakeGenericMethodEx(args);
+                      parameters = mi == null ? new ParameterInfo[0] : mi.GetParameters();
+                    }
+#endif
+                                if (args.Length != parameters.Length)
```

Comment:
Your indentation is all over the place. Do you should have replace tabs with places turned on and tabs/indents should be 4. It makes it very hard to read and review.